.github/workflows/test-go.yml

```diff
@@ -52,79 +52,103 @@ on:
-    #
-    # This job generates a JSON Array of integers ranging from 1 to 16.
-    # That array is used in the matrix section of the test-go job below.
-    #
-    outputs:
-      runner-indexes: ${{ steps.generate-index-list.outputs.indexes }}
     steps:
-    - id: generate-index-list
-      run: |
-        INDEX_LIST="$(seq 1 ${{ inputs.total-runners }})"
-        INDEX_JSON="$(jq --null-input --compact-output '. |= [inputs]' <<< "${INDEX_LIST}")"
-        echo "indexes=${INDEX_JSON}" >> "${GITHUB_OUTPUT}"
+      - uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # v3.5.3
+      - uses: actions/setup-go@fac708d6674e30b6ba41289acaab6d4b75aa0753 # v4.0.1
+        with:
+          go-version-file: ./.go-version
+          cache: true
+      - name: Authenticate to Vault
+        id: vault-auth
+        if: github.repository == 'hashicorp/vault-enterprise'
+        run: vault-auth
+      - name: Fetch Secrets
+        id: secrets
+        if: github.repository == 'hashicorp/vault-enterprise'
+        uses: hashicorp/vault-action@130d1f5f4fe645bb6c83e4225c04d64cfb62de6e
+        with:
+          url: ${{ steps.vault-auth.outputs.addr }}
+          caCertificate: ${{ steps.vault-auth.outputs.ca_certificate }}
+          token: ${{ steps.vault-auth.outputs.token }}
+          secrets: |
+            kv/data/github/${{ github.repository }}/datadog-ci DATADOG_API_KEY;
+            kv/data/github/${{ github.repository }}/github-token username-and-token | github-token;
+            kv/data/github/${{ github.repository }}/license license_1 | VAULT_LICENSE_CI;
+            kv/data/github/${{ github.repository }}/license license_2 | VAULT_LICENSE_2;
+            kv/data/github/${{ github.repository }}/hcp-link HCP_API_ADDRESS;
+            kv/data/github/${{ github.repository }}/hcp-link HCP_AUTH_URL;
+            kv/data/github/${{ github.repository }}/hcp-link HCP_CLIENT_ID;
+            kv/data/github/${{ github.repository }}/hcp-link HCP_CLIENT_SECRET;
+            kv/data/github/${{ github.repository }}/hcp-link HCP_RESOURCE_ID;
+      - id: setup-git-private
+        name: Setup Git configuration (private)
+        if: github.repository == 'hashicorp/vault-enterprise'
+        run: |
+          git config --global url."https://${{ steps.secrets.outputs.github-token }}@github.com".insteadOf https://github.com
+      - id: setup-git-public
+        name: Setup Git configuration (public)
+        if: github.repository != 'hashicorp/vault-enterprise'
+        run: |
+          git config --global url."https://${{ secrets.ELEVATED_GITHUB_TOKEN}}@github.com".insteadOf https://github.com
+      - run: go install gotest.tools/gotestsum@v1.9.0
+
+      - run: mkdir -p test-results/go-test
+      - id: restore-from-cache
+        uses: actions/cache@88522ab9f39a2ea568f7027eddc7d8d8bc9d59c8 # v3.3.1
+        with:
+          path: test-results/go-test
+          key: read-go-test-reports-${{ github.run_number }}
```

Comment:
what's the purpose of the cache key `read-go-test-reports-${{ github.run_number }}`? Is the intention that it won't match anything?

Reply:
I believe this is intended to pick up the last cached read (if it exists). Otherwise, it will fall back to the `test-results` cache populated at the end of a successful run using `restore-keys`: https://github.com/hashicorp/vault/pull/21745/files#diff-6aac4b916cfd203239c824402c61b8d7257d7c849d4730b81041f05ac4569515R287

I pulled it from the example here: https://github.com/dnephin/infra/blob/main/.github/workflows/ci-core.yaml#L22

Reply:
Ah ok. I don't think this key ever gets written to, since nothing else in the workflow has the key, and the [github.run_number](https://docs.github.com/en/actions/learn-github-actions/contexts) will be unique between runs. I guess since we have the restore-keys, it doesn't matter?

Reply:
I think it gets populated every time we restore (at the beginning of the current run) from `go-test-...`. I'm not sure if you can restore from cache with only `restore-keys`. I need to think about this a bit more. I'll get back to you -- sorry for the uninformed speculation.

Reply:
No worries, I also don't understand how it works 😅

Reply:
Okay, I've done a bit of investigation and this is my understanding:

You cannot do a restore from cache without a `key` specified, so we need a key.

We don't want to put dirty state in the cache, since we don't have good logs until the end of the test run.

We restore the cache to a temporary new cache with the `read-` prefix, which as you pointed out is unique for each test run.

At the end, we pull the data down and store it to the persistent cache, which should be matched using `restore-keys` above.

We _may_ want to delete the read cache at the end using the cache API, or we can just leave read caches in and let the eviction take care of it.

Reply:
might be worth adding a comment explaining why we use that key